hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java

```diff
@@ -25,74 +25,107 @@
-                          List<RingbufferMergeTypes> mergingEntries) {
+    public MergeOperation(ObjectNamespace namespace,
+                          SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> mergePolicy,
+                          Ringbuffer<Object> mergingRingbuffer) {
         this.namespace = namespace;
         this.mergePolicy = mergePolicy;
-        this.mergingEntries = mergingEntries;
+        this.mergingRingbuffer = mergingRingbuffer;
     }
 
     @Override
     public void beforeRun() throws Exception {
-        RingbufferService service = getService();
+        final RingbufferService service = getService();
         config = getRingbufferConfig(service, namespace);
-        ringbuffer = service.getOrCreateContainer(getPartitionId(), namespace, config);
     }
 
     @Override
     public void run() throws Exception {
-        valueMap = createHashMap(mergingEntries.size());
-        for (RingbufferMergeTypes mergingEntry : mergingEntries) {
-            long resultSequence = ringbuffer.merge(mergingEntry, mergePolicy);
-            if (resultSequence != -1) {
-                valueMap.put(resultSequence, ringbuffer.readAsData(resultSequence));
+        final RingbufferService service = getService();
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        RingbufferContainer<Object, Object> existingContainer = service.getContainerOrNull(getPartitionId(), namespace);
+
+        final RingbufferMergeTypes mergingValue = getMergingType(mergingRingbuffer);
+        final RingbufferMergeTypes existingValue =
+                existingContainer != null ? getMergingType(existingContainer.getRingbuffer()) : null;
+        serializationService.getManagedContext().initialize(mergePolicy);
+
+        final RingbufferMergeData<Object> resultData = mergePolicy.merge(mergingValue, existingValue);
+
+        if (resultData == null) {
+            service.destroyContainer(getPartitionId(), namespace);
+            resultRingbuffer = null;
+        } else {
+            if (existingContainer == null) {
+                final RingbufferConfig config = getRingbufferConfig(service, namespace);
+                existingContainer = service.getOrCreateContainer(getPartitionId(), namespace, config);
             }
+            resultRingbuffer = new ArrayRingbuffer<Object>(resultData);
+            existingContainer.setRingbuffer(resultRingbuffer);
         }
     }
 
+    /**
+     * Returns the {@link RingbufferMergeTypes} of the provided ringbuffer that
+     * can be provided to the merge policy.
+     *
+     * @param ringbuffer the ringbuffer containing the merge data
+     * @return the merge type corresponding to the given ringbuffer
+     */
+    private RingbufferMergeTypes getMergingType(Ringbuffer<Object> ringbuffer) {
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
```

Comment:
You could pass the `SerializationService` as parameter, so you don't have to retrieve it twice.